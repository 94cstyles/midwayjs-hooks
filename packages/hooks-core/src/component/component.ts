import isFunction from 'lodash/isFunction'
import noop from 'lodash/noop'
import pickBy from 'lodash/pickBy'
import upath from 'upath'

import { createConfiguration, IMidwayApplication } from '@midwayjs/core'

import { lazyRequire, Route } from '..'
import { EXPORT_DEFAULT_FUNCTION_ALIAS } from '../const'
import { GatewayManager } from '../gateway/manager'
import { ApiFunction, ApiModule } from '../types/common'
import { ComponentOptions, HooksGatewayAdapter } from '../types/gateway'
import { useHooksMiddleware } from '../util'

export type LoadApiModuleOption = {
  mod: ApiModule
  file: string
}

export class HooksComponent {
  options: ComponentOptions
  gatewayManager: GatewayManager

  constructor(options: ComponentOptions) {
    this.options = options
    this.gatewayManager = GatewayManager.getInstance(
      this.options.root,
      this.options.projectConfig
    )
  }

  load() {
    this.loadByScanner()

    const configuration = createConfiguration({
      namespace: '@midwayjs/hooks',
    })

    configuration
      .onReady(async (container, app: IMidwayApplication) => {
        for (const gateway of this.gatewayManager.gateways) {
          await gateway.onReady?.({
            app,
            runtimeConfig: this.options.runtimeConfig,
          })
        }
      })
      .onStop(noop)

    return {
      Configuration: configuration,
    }
  }

  loadByScanner() {
    const { sync } = lazyRequire<typeof import('globby')>('globby')

    const {
      router,
      projectConfig: { routes },
    } = this.options

    for (const route of routes) {
      // Windows
      const dir = upath.join(router.source, route.baseDir)
      const files = sync([dir]).filter((file) => router.isApiFile(file))

      for (const file of files) {
        const mod: ApiModule = require(file)
        this.loadApiModule({ mod, file })
      }
    }
  }

  loadApiModule({ mod, file }: LoadApiModuleOption) {
    const route = this.options.router.getRoute(file)
    const gateway = this.gatewayManager.getGatewayByRoute(route)
    this.createApi(mod, gateway, file, route)
  }

  createApi(
    mod: ApiModule,
    gateway: HooksGatewayAdapter,
    file: string,
    route: Route
  ) {
    const modMiddleware = mod?.config?.middleware || []
    const funcs = pickBy<ApiFunction>(mod, isFunction)

    for (const [name, fn] of Object.entries(funcs)) {
      fn.middleware = (fn.middleware || (fn.middleware = []))
        .concat(modMiddleware)
        .map(useHooksMiddleware)

      const isExportDefault = name === 'default'
      const functionName = isExportDefault
        ? EXPORT_DEFAULT_FUNCTION_ALIAS
        : name
      const functionId = this.options.router.getFunctionId(
        file,
        functionName,
        isExportDefault
      )

      fn._param = { functionId }

      gateway.createApi({
        fn,
        functionName,
        isExportDefault,
        functionId,
        file,
        route,
      })
    }
  }
}
