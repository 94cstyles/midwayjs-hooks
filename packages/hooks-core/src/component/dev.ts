import { createConfiguration, IMidwayContainer } from '@midwayjs/core'
import { __decorate } from 'tslib'
import { Inject, Controller, Get, Post, Provide } from '@midwayjs/decorator'
import { als } from '../runtime'
import { ApiFunction, ApiModule } from '../types/common'
import { ServerRouter } from '../router'
import { RuntimeConfig, InternalConfig } from '../types/config'
import { kebabCase, noop } from 'lodash'
import { extname, relative, removeExt } from 'upath'
import { ApiHttpMethod } from '../types/http'
import { superjson } from '../lib'
import { join } from 'path'
import { existsSync } from 'fs'

export type ComponentConfig = {
  runtime: RuntimeConfig
  internal: InternalConfig
  router: ServerRouter
  root: string
}

export class HooksDevComponent {
  internal: InternalConfig
  runtime: RuntimeConfig
  router: ServerRouter
  root: string
  container: IMidwayContainer

  constructor(config: ComponentConfig) {
    this.runtime = config.runtime
    this.internal = config.internal
    this.router = config.router
    this.root = config.root
  }

  init() {
    let count = 0
    const { routes } = this.internal
    const configuration = createConfiguration({
      namespace: '@midwayjs/hooks',
      directoryResolveFilter: routes.map((route) => {
        return {
          pattern: route.baseDir,
          ignoreRequire: true,
          filter: (_: void, file: string, container: IMidwayContainer) => {
            if (!this.router.isApiFile(file)) {
              return
            }
            this.container = container
            this.createApi(file)

            count++
            if (count === routes.length) {
              this.afterCreate?.()
            }
          },
        }
      }),
    })

    configuration
      .onReady((_, app) => {
        this.useGlobalMiddleware(app)
      })
      .onStop(noop)

    return {
      Configuration: configuration,
    }
  }

  afterCreate() {}

  createApi(file: string) {
    const mod: ApiModule = require(file)
    const modMiddleware = mod?.config?.middleware || []

    Object.keys(mod)
      .filter((key) => typeof mod[key] === 'function')
      .forEach((key) => {
        this.createApiFunction({
          fn: mod[key],
          file: file,
          isExportDefault: key === 'default',
          modMiddleware,
        })
      })
  }

  createApiFunction(config: {
    fn: ApiFunction
    file: string
    isExportDefault: boolean
    modMiddleware: any[]
  }) {
    const { fn, file, isExportDefault, modMiddleware } = config

    const fnName = isExportDefault ? '$default' : fn.name
    const id = this.getFunctionId(file, fnName, isExportDefault)

    const containerId = 'hooks::' + id
    const httpPath = this.router.getHTTPPath(file, fnName, isExportDefault)
    const httpMethod: ApiHttpMethod = fn.length === 0 ? 'GET' : 'POST'

    // Set param for unit testing
    fn._param = {
      url: httpPath,
      method: httpMethod,
      meta: { functionName: id },
    }

    // Apply mod middleware
    fn.middleware = (fn.middleware || (fn.middleware = []))
      .concat(modMiddleware)
      .map(this.useHooksMiddleware)

    this.registerApiFunction({
      containerId,
      httpMethod,
      httpPath,
      fn,
    })
  }

  getFunctionId(file: string, functionName: string, isExportDefault: boolean) {
    const relativePath = relative(this.router.source, file)
    // src/apis/lambda/index.ts -> apis-lambda-index
    const id = kebabCase(removeExt(relativePath, extname(relativePath)))
    const name = [id, isExportDefault ? '' : `-${functionName}`].join('')
    return name.toLowerCase()
  }

  registerApiFunction(config: {
    containerId: string
    httpMethod: ApiHttpMethod
    httpPath: string
    fn: ApiFunction
  }) {
    const { containerId, httpMethod, httpPath, fn } = config
    const Method = httpMethod === 'GET' ? Get : Post
    const enableSuperjson = this.internal.superjson

    let FunctionContainer = class FunctionContainer {
      ctx: any
      async handler() {
        let args = this.ctx.request?.body?.args || []
        if (typeof args === 'string') {
          args = JSON.parse(args)
        }

        const response = await fn(...args)
        if (enableSuperjson) {
          return superjson.serialize(response)
        }
        return response
      }
    }
    __decorate([Inject()], FunctionContainer.prototype, 'ctx', void 0)
    __decorate(
      [Method(httpPath, { middleware: fn.middleware })],
      FunctionContainer.prototype,
      'handler',
      null
    )
    FunctionContainer = __decorate(
      [Provide(containerId), Controller('/')],
      FunctionContainer
    )
    this.container.bind(containerId, FunctionContainer)
  }

  useGlobalMiddleware(app: any) {
    app.use(this.useAsyncLocalStorage)

    // Apply global middleware from config
    if (Array.isArray(this.runtime.middleware)) {
      this.runtime.middleware.forEach((middleware) =>
        app.use(this.useHooksMiddleware(middleware))
      )
    }
  }

  useAsyncLocalStorage = async (ctx: any, next: any) => {
    const enableSuperjson = this.internal.superjson
    await als.run({ ctx }, async () => {
      try {
        await next()
      } catch (error) {
        if (enableSuperjson) {
          ctx.status = 500
          ctx.body = superjson.serialize(error)
        } else {
          throw error
        }
      }
    })
  }

  useHooksMiddleware(fn: Function) {
    return (...args: any[]) => {
      /**
       * Hooks middleware
       * const middleware = (next) => { const ctx = useContext() }
       */
      if (fn.length === 1) {
        const next = args[args.length - 1]
        return fn(next)
      }
      return fn(...args)
    }
  }

  isFullStackProject() {
    return [
      // Vite
      'vite.config.ts',
      'vite.config.js',
      // build-scritps
      'build.json',
      'build.js',
    ].some((config) => existsSync(join(this.root, config)))
  }
}
