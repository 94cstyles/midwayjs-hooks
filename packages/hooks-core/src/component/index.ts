import {
  createConfiguration,
  IMidwayContainer,
  MidwayFrameworkType,
} from '@midwayjs/core'
import { __decorate } from 'tslib'
import { Inject, Controller, Get, Post, Provide } from '@midwayjs/decorator'
import { als } from '../runtime'
import { ApiFunction, ApiModule } from '../types/common'
import { ServerRouter } from '../router'
import { getConfig, getProjectRoot } from '../config'
import { InternalConfig } from '../types/config'
import { isProduction } from '../util'
import { kebabCase, noop } from 'lodash'
import { join } from 'path'
import staticCache from 'koa-static-cache'
import { extname, relative } from 'upath'

/**
 * Create hooks component
 */
export const hooks = () => {
  return new HooksComponent().createConfiguration()
}

class HooksComponent {
  private readonly root: string
  private readonly config: InternalConfig
  private readonly router: ServerRouter
  private container: IMidwayContainer

  constructor() {
    this.root = getProjectRoot()
    this.config = getConfig()
    this.router = new ServerRouter(this.root, this.config)
  }

  createConfiguration() {
    const configuration = createConfiguration({
      namespace: '@midwayjs/hooks',
      directoryResolveFilter: this.config.routes.map((route, index) => {
        return {
          pattern: route.baseDir,
          ignoreRequire: true,
          filter: (
            mod: void,
            sourceFilePath: string,
            container: IMidwayContainer
          ) => {
            if (!this.container) this.container = container
            if (!this.router.isApiFile(sourceFilePath)) return

            this.createApiFromSourceFile(sourceFilePath)

            if (index === this.config.routes.length - 1) {
              this.createRenderFunction()
            }
          },
        }
      }),
    })

    configuration
      .onReady((container, app) => {
        this.applyMiddleware(app)
      })
      .onStop(noop)

    return {
      Configuration: configuration,
    }
  }

  createApiFromSourceFile(sourceFilePath: string) {
    const mod: ApiModule = require(sourceFilePath)
    const modMiddleware = mod?.config?.middleware ? mod.config.middleware : []

    if (typeof mod === 'function') {
      this.createFunction({
        fn: mod,
        sourceFilePath,
        isExportDefault: true,
        modMiddleware,
      })
    }

    Object.keys(mod)
      .filter((key) => typeof mod[key] === 'function')
      .forEach((key) => {
        this.createFunction({
          fn: mod[key],
          sourceFilePath,
          isExportDefault: key === 'default',
          modMiddleware,
        })
      })
  }

  private createFunction(config: {
    fn: ApiFunction
    sourceFilePath: string
    isExportDefault: boolean
    modMiddleware: any[]
  }) {
    const { fn, sourceFilePath, isExportDefault, modMiddleware } = config

    const fnName = isExportDefault ? '$default' : fn.name
    const id = this.getFunctionId({
      sourceFilePath,
      functionName: fnName,
      isExportDefault,
    })

    const containerId = 'hooks::' + id
    const httpPath = this.router.getHTTPPath(
      sourceFilePath,
      fnName,
      isExportDefault
    )
    const httpMethod = fn.length === 0 ? 'GET' : 'POST'

    // Set param for unit testing
    fn._param = {
      url: httpPath,
      method: httpMethod,
      meta: { functionName: id },
    }

    // Apply module middleware
    ;(fn.middleware || (fn.middleware = [])).unshift(...modMiddleware)

    this.registerFunctionToContainer({
      containerId,
      httpMethod,
      httpPath,
      fn,
    })
  }

  private registerFunctionToContainer(config: {
    containerId: string
    httpMethod: any
    httpPath: string
    fn: ApiFunction
  }) {
    const { containerId, httpMethod, httpPath, fn } = config
    const Method = httpMethod === 'GET' ? Get : Post

    // Source: https://shorturl.at/pqI06
    let FunctionContainer = class FunctionContainer {
      ctx: any
      async handler() {
        const bindCtx = { ctx: this.ctx }
        let args = this.ctx.request?.body?.args || []
        if (typeof args === 'string') {
          args = JSON.parse(args)
        }
        return await als.run(bindCtx, async () => fn(...args))
      }
    }
    __decorate([Inject()], FunctionContainer.prototype, 'ctx', void 0)
    __decorate(
      [Method(httpPath, { middleware: fn.middleware || [] })],
      FunctionContainer.prototype,
      'handler',
      null
    )
    FunctionContainer = __decorate(
      [Provide(containerId), Controller('/')],
      FunctionContainer
    )
    this.container.bind(containerId, FunctionContainer)
  }

  private hasRender = false
  private createRenderFunction() {
    if (!isProduction() || this.hasRender) {
      return
    }

    const fn = async () => {}
    this.registerFunctionToContainer({
      containerId: 'hooks:page-render',
      httpMethod: 'GET',
      httpPath: '/*',
      fn,
    })

    this.hasRender = true
  }

  private applyMiddleware(app: any) {
    // Apply global middleware from config
    this.config.middleware.forEach((middleware) => app.use(middleware))

    // Serve vite static html
    const type = app.getFrameworkType()
    const requireStaticCache =
      type === MidwayFrameworkType.WEB_KOA || type === MidwayFrameworkType.FAAS

    if (isProduction() && requireStaticCache) {
      const baseDir = app.getBaseDir()
      app.use(
        staticCache({
          dir: join(baseDir, '..', this.config.build.viteOutDir),
          dynamic: true,
          alias: {
            '/': 'index.html',
          },
          buffer: true,
        })
      )
    }
  }

  private getFunctionId(config: {
    sourceFilePath: string
    functionName: string
    isExportDefault: boolean
  }) {
    const { sourceFilePath, functionName, isExportDefault } = config

    const rule = this.router.getRouteConfigBySourceFilePath(sourceFilePath)
    const lambdaDirectory = this.router.getApiDirectory(rule.baseDir)

    const length = this.router.config.routes.length
    // 多个 source 的情况下，根据各自的 lambdaDirectory 来增加前缀命名
    const relativeDirectory = length > 1 ? this.router.source : lambdaDirectory
    const relativePath = relative(relativeDirectory, sourceFilePath)
    // a/b/c -> a-b-c
    const id = kebabCase(removeExtension(relativePath))
    const name = [id, isExportDefault ? '' : `-${functionName}`].join('')
    return name.toLowerCase()
  }
}

export function removeExtension(file: string) {
  return file.replace(extname(file), '')
}
