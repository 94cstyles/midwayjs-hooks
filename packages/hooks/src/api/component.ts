import { createConfiguration, IMidwayApplication } from '@midwayjs/core'
import {
  RuntimeConfig,
  validateArray,
  als,
  useHooksMiddleware,
  HooksMiddleware,
} from '@midwayjs/hooks-core'

interface MidwayApplication extends IMidwayApplication {
  use?: (middleware: any) => void
}

export function HooksComponent(runtimeConfig: RuntimeConfig = {}) {
  if (runtimeConfig.middleware !== undefined) {
    validateArray(runtimeConfig.middleware, 'runtimeConfig.middleware')
  }

  const Configuration = createConfiguration({
    namespace: '@midwayjs/hooks',
  }).onReady((_, app: MidwayApplication) => {
    registerGlobalMiddleware(app, runtimeConfig.middleware)
  })

  // registry api routes

  return { Configuration }
}

function registerGlobalMiddleware(
  app: MidwayApplication,
  middlewares: HooksMiddleware[] = []
) {
  app.use?.(useHooksRuntime)
  for (const mw of middlewares) {
    app.use?.(useHooksMiddleware(mw))
  }
}

// For http
async function useHooksRuntime(ctx: any, next: any) {
  await als.run({ ctx }, async () => await next())
}
