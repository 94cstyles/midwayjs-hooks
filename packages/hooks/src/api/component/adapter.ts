import {
  AbstractFrameworkAdapter,
  AbstractRouter,
  ContextManager,
  ApiRoute,
  createDebug,
  HooksMiddleware,
  HttpTrigger,
  HttpTriggerType,
  isHooksMiddleware,
  ResponseMetaData,
  ResponseMetaType,
  urlJoin,
  useContext,
  validateOneOf,
} from '@midwayjs/hooks-core'
import { IMidwayApplication, IMidwayContainer } from '@midwayjs/core'
import {
  All,
  Controller,
  Del,
  Get,
  Head,
  MidwayFrameworkType,
  Options,
  Patch,
  Post,
  Put,
  ServerlessTrigger,
  ServerlessTriggerType,
} from '@midwayjs/decorator'
import {
  createFunctionContainer,
  isDev,
  isFileSystemRouter,
} from '../../internal'
import {
  HSFTrigger,
  MTopTrigger,
  ServerlessTimerTrigger,
} from '../operator/serverless'

const debug = createDebug('hooks: MidwayFrameworkAdapter')

export interface MidwayApplication extends IMidwayApplication {
  use?: (middleware: any) => void
}

export class MidwayFrameworkAdapter extends AbstractFrameworkAdapter {
  constructor(
    router: AbstractRouter,
    public app: MidwayApplication,
    public container: IMidwayContainer
  ) {
    super(router)
  }

  private get frameworkType() {
    return this.app.getFrameworkType()
  }

  private controllers = []
  bindControllers() {
    for (const controller of this.controllers) {
      this.container.bind(controller)
    }
  }

  isHttpTrigger(api: ApiRoute): api is ApiRoute<HttpTrigger> {
    return api.trigger.type === 'HTTP'
  }

  isServerlessTimerTrigger(
    api: ApiRoute
  ): api is ApiRoute<ServerlessTimerTrigger> {
    return api.trigger.type === ServerlessTriggerType.TIMER
  }

  isMTopTrigger(api: ApiRoute): api is ApiRoute<MTopTrigger> {
    return api.trigger.type === ServerlessTriggerType.MTOP
  }

  isHSFTrigger(api: ApiRoute): api is ApiRoute<HSFTrigger> {
    return api.trigger.type === ServerlessTriggerType.HSF
  }

  registerApiRoutes(apis: ApiRoute[]) {
    for (const api of apis) {
      if (this.isHttpTrigger(api)) {
        api.middleware = api.middleware?.map((mw) =>
          this.useHooksMiddleware(mw)
        )
        this.controllers.push(this.createHttpApi(api))
        continue
      }

      if (this.isHSFTrigger(api)) {
        this.controllers.push(this.createHSFApi(api))
        continue
      }

      if (this.isMTopTrigger(api)) {
        this.controllers.push(this.createMTopApi(api))
        continue
      }

      if (this.isServerlessTimerTrigger(api)) {
        this.controllers.push(this.createServerlessTimer(api))
        continue
      }

      throw new Error(`Unsupported trigger type: ${api.trigger.type}`)
    }
  }

  createHSFApi(api: ApiRoute<HSFTrigger>) {
    const { functionId, fn } = api

    debug('create hsf api: %s', functionId)

    return createFunctionContainer({
      fn,
      functionId,
      parseArgs({ args }) {
        const event = args[0]
        return event?.args || []
      },
      handlerDecorators: [ServerlessTrigger(ServerlessTriggerType.HSF)],
    })
  }

  createMTopApi(api: ApiRoute<MTopTrigger>) {
    const { functionId, fn } = api

    debug('create mtop api: %s', functionId)

    return createFunctionContainer({
      fn,
      functionId,
      parseArgs({ args }) {
        const event = args[0]
        return event?.args || []
      },
      handlerDecorators: [ServerlessTrigger(ServerlessTriggerType.MTOP)],
    })
  }

  createServerlessTimer(api: ApiRoute<ServerlessTimerTrigger>) {
    const { trigger, fn, functionId } = api

    debug('create serverless timer: %s %O', functionId, trigger.options)

    return createFunctionContainer({
      fn,
      functionId,
      parseArgs({ args }) {
        return args
      },
      handlerDecorators: [
        ServerlessTrigger(ServerlessTriggerType.TIMER, trigger.options),
      ],
    })
  }

  createHttpApi(api: ApiRoute) {
    const { functionId, fn, trigger } = api

    validateOneOf(
      trigger.method,
      'trigger.method',
      Object.keys(this.methodDecorators)
    )
    const Method = this.methodDecorators[trigger.method]
    const url = normalizeUrl(this.router, api)

    debug('create http api: %s %s %s', functionId, trigger.method, url)

    if (isDev()) {
      globalThis['HOOKS_ROUTER'] ??= []
      globalThis['HOOKS_ROUTER'].push({
        type: HttpTriggerType.toLowerCase(),
        path: url,
        method: trigger.method,
        functionId,
        handler: `${functionId}.handler`,
      })
    }

    return createFunctionContainer({
      fn,
      functionId,
      parseArgs({ ctx }) {
        return ctx.request?.body?.args || []
      },
      classDecorators: [Controller()],
      handlerDecorators: [Method(url, { middleware: api.middleware })],
    })
  }

  private methodDecorators = {
    GET: Get,
    POST: Post,
    PUT: Put,
    DELETE: Del,
    PATCH: Patch,
    HEAD: Head,
    OPTIONS: Options,
    ALL: All,
  }

  registerGlobalMiddleware(middlewares: HooksMiddleware[] = []) {
    const runtime =
      this.frameworkType === MidwayFrameworkType.WEB_EXPRESS
        ? this.useExpressRuntime
        : this.useUniversalRuntime

    this.app.use?.(runtime)
    for (const mw of middlewares) {
      this.app.use?.(this.useHooksMiddleware(mw))
    }
  }

  private async useExpressRuntime(req: any, res: any, next: any) {
    throw new Error('Express runtime is not supported. Please use koa.')
  }

  private async useUniversalRuntime(ctx: any, next: any) {
    await ContextManager.run({ ctx }, async () => await next())
  }

  private useHooksMiddleware(fn: (...args: any[]) => any) {
    if (!isHooksMiddleware(fn)) return fn

    return (...args: any[]) => {
      const next =
        this.frameworkType === MidwayFrameworkType.WEB_EXPRESS
          ? args[args.length - 1]
          : args[1]
      return fn(next)
    }
  }

  async handleResponseMetaData(metadata: ResponseMetaData[]): Promise<any> {
    const ctx = useContext()

    for (const meta of metadata) {
      switch (meta.type) {
        case ResponseMetaType.CODE:
          ctx.status = meta.code
          break
        case ResponseMetaType.HEADER:
          ctx.set(meta.header.key, meta.header.value)
          break
        case ResponseMetaType.CONTENT_TYPE:
          ctx.type = meta.contentType
          break
        case ResponseMetaType.REDIRECT:
          ctx.status = meta.code || 302
          ctx.redirect(meta.url)
          break
      }
    }
  }
}

export function normalizeUrl(router: AbstractRouter, api: ApiRoute) {
  const { trigger, file } = api

  if (isFileSystemRouter(router)) {
    const basePath = router.getRoute(file).basePath
    return urlJoin(basePath, trigger.path, {})
  }

  return trigger.path
}
