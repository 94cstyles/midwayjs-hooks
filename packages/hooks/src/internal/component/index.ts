import { FileRouter, validateArray, isDevelopment } from '@midwayjs/hooks-core'

import { getConfig, getProjectRoot } from '../config'
import { RuntimeConfig } from '../config/type'
import { HooksComponent } from './component'

/**
 * @description Create hooks component
 */
export const hooks = (runtimeConfig: RuntimeConfig = {}) => {
  if (runtimeConfig.middleware !== undefined) {
    validateArray(runtimeConfig.middleware, 'runtimeConfig.middleware')
  }

  const root = getProjectRoot()
  const projectConfig = getConfig()

  const router = new FileRouter({
    root: root,
    projectConfig: projectConfig,
    useSourceFile: isDevelopment(),
  })

  const hooks = new HooksComponent({
    root: root,
    router,
    runtimeConfig,
    projectConfig: projectConfig,
  })

  return hooks.load()
}
