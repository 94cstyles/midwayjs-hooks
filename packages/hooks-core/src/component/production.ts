import { HooksDevComponent } from './dev'
import staticCache from 'koa-static-cache'
import { join } from 'upath'
import { consola } from '../lib'

export class HooksProductionComponent extends HooksDevComponent {
  useGlobalMiddleware(app: any) {
    super.useGlobalMiddleware(app)

    if (!this.isFullStackProject()) {
      return
    }

    if (this.existRootPath) {
      return
    }

    const baseDir = app.getBaseDir()
    const dist = join(baseDir, '..', this.internal.build.viteOutDir)
    const mw = staticCache({
      dir: dist,
      dynamic: true,
      alias: {
        '/': 'index.html',
      },
      buffer: true,
      gzip: true,
    })
    app.use(mw)
  }

  afterCreate() {
    if (!this.isFullStackProject()) {
      return
    }

    if (this.existRootPath) {
      consola.debug(
        'The route `/` or `/*` is registered, you need to host the front-end page manually'
      )
      return
    }

    this.registerApiFunction({
      containerId: 'hooks:host',
      httpMethod: 'GET',
      httpPath: '/*',
      fn: async () => {},
    })
  }

  get existRootPath() {
    const { routes } = this.router
    return routes.has('/') || routes.has('/*')
  }
}
