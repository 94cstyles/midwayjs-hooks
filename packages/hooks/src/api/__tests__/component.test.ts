import { HooksComponent } from '../component'

test('component should validate arguments', () => {
  expect(() =>
    HooksComponent({ middleware: null })
  ).toThrowErrorMatchingSnapshot()
})
