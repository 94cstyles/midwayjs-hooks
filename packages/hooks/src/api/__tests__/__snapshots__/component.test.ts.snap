// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component should validate arguments 1`] = `"[ERR_INVALID_ARG_TYPE]: The 'runtimeConfig.middleware' argument must be of type Array. Received null"`;
